### PR TITLE
Explain the use of iterative solvers in the introduction of step-3.

### PR DESCRIPTION
**Pull request by bangerth**
The introduction of step-3 is in need of some expansion. One of the missing pieces is that it never says anything about how we solve linear systems.

Found while thinking about #9596.

### CODE REVIEW

File: examples/step-3/doc/intro.dox [hunk 1]
```diff
@@ -149,7 +149,7 @@ takes to make that happen:
   linear systems.
 - We need a way to form the integrals. In the finite element method, this is
   most commonly done using quadrature, i.e. the integrals are replaced by a
-  weighted sum over a set of points on each cell. That is, we first split the
+  weighted sum over a set of "quadrature points" on each cell. That is, we first split the
```
Comment (drwells) [inline examples/step-3/doc/intro.dox:152]:
```suggestion
  weighted sum over a set of *quadrature points* on each cell. That is, we first split the
```
The other technical terms (like assembly) are emphasized so lets do that too.

File: examples/step-3/doc/intro.dox [hunk 2]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+and (ii) it only ever requires matrix-vector products, which is very useful
+for sparse matrices because a sparse matrix has, by definition, only
+${\cal O}(N)$ entries and so a matrix-vector product can be done with
+${\cal O}(N)$ effort whereas it costs $N^2$ operations to do the same for
+dense matrices. As a consequence, we can hope to solve linear systems
+with at most ${\cal O}(N^2)$ operations, and in some cases substantially
+fewer.
+
+Finite element codes therefore almost always use iterative solvers such as
+Conjugate Gradients for the solution of the linear systems, and we will
```
Comment (drwells) [inline examples/step-3/doc/intro.dox:303]:
```suggestion
CG for the solution of the linear systems, and we will
```
we called it 'Conjugate Gradient' earlier (singular).

File: examples/step-3/doc/intro.dox [hunk 3]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+for matrices that are symmetric and positive definite; for other equations,
+the matrix may not have these properties and we will have to use other
+variations of iterative solvers such as 
+[BiCGStab](https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method)
+or [GMRES](https://en.wikipedia.org/wiki/Generalized_minimal_residual_method)
+that are applicable to more general matrices.)
+
+An important component of these iterative solvers is that we specify
+the *tolerance* with which we want to solve the linear system -- in essence,
+a statement about the error we are willing to accept in our approximation
```
Comment (drwells) [inline examples/step-3/doc/intro.dox:314]:
```suggestion
a statement about the error we are willing to accept in our approximate
```
again, just a consistency change

File: examples/step-3/doc/intro.dox [hunk 4]
```diff
@@ -138,7 +138,7 @@ tutorial where we have a non-symmetric bilinear form for which it makes a
 difference whether we multiply from the right or from the left.
 
 
-<h3> Computing the matrix and right hand side vector </h3>
+<h3> "Assembling" the matrix and right hand side vector </h3>
```
Comment (drwells) [inline examples/step-3/doc/intro.dox:141]:
```suggestion
<h3> *Assembling* the matrix and right hand side vector </h3>
```

**Review by drwells — CHANGES_REQUESTED** [review]
Looks good. I have some suggestions regarding consistenty of names and formatting.

**Review by drwells — APPROVED** [review]

**Review by tjhei — APPROVED** [review]
Very nice!👍

File: examples/step-3/doc/intro.dox [hunk 5]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+faster; we will use one such method -- implemented in the SparseDirectUMFPACK
+class -- in step-29 for the first time, among several others than come after
+that. These variations of Gaussian elimination might get us to problem
+sizes on the order of 100,000 or 200,000, but not all that much beyond that.
+
+Instead, what we will do here is take up an idea from 1952: the
+[Conjugate Gradient method](https://en.wikipedia.org/wiki/Conjugate_gradient_method),
+or in short "CG". CG is an "iterative" solver in that it forms a sequence
+of vectors that *converge* to the exact solution; in fact, after $N$ such iterations
+it finds the exact solution if the matrix is symmetric and positive definite.
```
Comment (kronbichler) [inline examples/step-3/doc/intro.dox:283]:
```suggestion
in the absence of roundoff errors it finds the exact solution if the matrix is symmetric and positive definite.
```

File: examples/step-3/doc/intro.dox [hunk 6]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+The method was originally developed as another way
+to solve a linear system exactly, like Gaussian elimination, but as such
+it had few advantages and was largely forgotten for a few decades. But,
```
Comment (kronbichler) [inline examples/step-3/doc/intro.dox:286]: If I recall the historic texts correctly, one of the main reasons was indeed the much higher sensitivity to roundoff errors than Gaussian elimination (today I would also mention data locality, but that was not an issue initially) - already for moderately ill-conditioned matrices (like `diag([1,2,3,...,100])` you can really see this happen.

File: examples/step-3/doc/intro.dox [hunk 7]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+when computers became powerful enough to solve problems of a size where
+Gaussian elimination doesn't work well any more (sometime in the 1980s),
+CG was rediscovered as a way to people realized that it is well suited for
```
Comment (kronbichler) [inline examples/step-3/doc/intro.dox:289]: I don't quite get this. Do you mean something like
```suggestion
CG was rediscovered as people realized that it is well suited for
```

File: examples/step-3/doc/intro.dox [hunk 8]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+large and sparse systems like the ones we get from the finite element method.
+This is because (i) the vectors it computes *converge* to the exact solution,
+and consequently we do not actually have to do all $N$ iterations to
+find the exact solution as long as we're happy with reasonably good approximations;
+and (ii) it only ever requires matrix-vector products, which is very useful
+for sparse matrices because a sparse matrix has, by definition, only
+${\cal O}(N)$ entries and so a matrix-vector product can be done with
+${\cal O}(N)$ effort whereas it costs $N^2$ operations to do the same for
+dense matrices. As a consequence, we can hope to solve linear systems
+with at most ${\cal O}(N^2)$ operations, and in some cases substantially
```
Comment (kronbichler) [inline examples/step-3/doc/intro.dox:299]:
```suggestion
with at most ${\cal O}(N^2)$ operations, and in many cases substantially
```
Poisson-type problems (which is one of the few cases where plain CG actually works) have a cost of O(N^{3/2}) in 2D and O(N^{4/3}) in 3D. The reason for the different powers is that in 3D, for the same N ~ h^{-3}, the ratio between the domain size and cell length is lower.

**Review by kronbichler — COMMENTED** [review]
A few comments regarding computing time and CG (not all comments are requests for change, some are merely comments).

File: examples/step-3/doc/intro.dox [hunk 9]
```diff
@@ -231,6 +238,105 @@ page. An overview of the most fundamental groups of concepts is also available
...
+linear systems.
+
+The first method one typically learns for solving linear systems is
+[Gaussian elimination](https://en.wikipedia.org/wiki/Gaussian_elimination).
+The problem with this method is that it requires a number of operations that
+is proportional to $N^3$, where $N$ is the number of equations or unknowns
+in the linear system -- more specifically, the number of operations is
+$\frac 23 N^3$, give or take a few. With $N=1089$, this means that we would
+have to do around $861$ million operations. This is a number that is quite
+feasible and it would take modern processors on the order of 1-10 seconds
```
[the remark this answers is not in the document]
Comment (bangerth) [reply examples/step-3/doc/intro.dox]: Ah, interesting. In my head, I thought that 861M operations will each take ~3 cycles, which gives us about a second. Then I padded the number by a safety factor for bandwith limitations. But thinking about it, a 1089x1089 matrix comfortably fits into the cache these days, and I think what you're saying is that with XMM, SSE, AVX, ... you get enough parallelism that things are actually that much faster. Fun.

File: examples/step-3/doc/intro.dox (hunk 6 above)
Comment (bangerth) [reply examples/step-3/doc/intro.dox:286]: Yes, round-off is a concern, but that's something that I believe people knew how to deal with -- with iterative refinement. Of course, round-off was a concern also for direct solver, given that they only had single-precision at the time. Basically, the whole issue of forward and backward stability analysis is a product of having only single precision available.

One could expand on all of this, but in the end, that isn't the place for it. I just wanted to briefly touch on the basic motivation to not use direct solvers in FEM codes, and so saving the historical context for another place might be the place to go :-) I don't think you suggested anything else either.

**Comment by bangerth** [comment]
So modified.

File: examples/step-3/doc/intro.dox (hunk 8 above)
Comment (bangerth) [reply examples/step-3/doc/intro.dox]: I assume you were talking about direct solvers:
Just for completeness, I think it's O(N^{5/3}) in 3d -- worse than the 3/2 in 2d. In 2d, you can re-sort DoFs in such a way that the bandwidth of the matrix is b=O(sqrt(N)), but in 3d the bandwidth becomes b=O(N^{2/3}).

File: examples/step-3/doc/intro.dox (hunk 6 above)
Comment (kronbichler) [reply examples/step-3/doc/intro.dox:286]: Exactly, I didn't want to suggest anything else here, I just wanted to add some thoughts.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**